### ISSUE TITLE
TMDB episode groups

### ISSUE THREAD

**Issue by hcgiub001** [Tikipeter/tikipeter.github.io#114]
some shows don't have correct season and episode listings
https://www.reddit.com/r/Addons4Kodi/comments/1ela4bt/does_fenfen_lite_or_umbrella_provides_latest/

it's not just related to anime, you have netflix shows to
https://www.reddit.com/r/Addons4Kodi/comments/1f0dt9m/tmdb_season_metadata_option_for_shows_in_fen/

I think I found the solution. TMDB episode groups
https://developer.themoviedb.org/reference/tv-series-episode-groups

since data is added by users on TMDB, all anime on TMDB has basically all episode groups

so for one piece which is so popular you have

![Screenshot (505)](https://github.com/user-attachments/assets/a457dd0e-8639-4835-b891-19c299778149)

right now the one piece episode group it's using has the last season as this
![Screenshot (507)](https://github.com/user-attachments/assets/2f80a79b-068e-471a-be10-acf679134744)

it's asking for season 22 and episode number in the thousands, but no links are being found. 
I did a manual scrape for season 22 and used the "seasons original air date" episode group and so I searched for episode 31 (instead of the equivalent absolute number of the episode) and had lots of results which is expected since it's the number 1 anime.

same for the other show that the person had issues finding
that show is showing up as 1 season 16 episodes,
but the torrents uploaded are catered for netflix episode groups that has 2 seasons 8 episodes each
so when Fenlight looks for season 1 episode 9 or above, no results will be matched. But if we can use alternative season groups , it'll match so before

![Screenshot (509)](https://github.com/user-attachments/assets/54c4edcc-5425-4cb0-be38-ebbf65fb6c53)

and now
![Screenshot (511)](https://github.com/user-attachments/assets/d203594f-5c3f-4060-9bde-728a7aa69b4e)

I was thinking if we want to change episode groups we could have it in context menu on the show. then when someone clicks on alternative season episode groups you can pull the list like so
![Screenshot (512)](https://github.com/user-attachments/assets/d79be94a-1ed9-41ab-abb2-c1bb32172919)

If you can chose which group to have as default that would be awesome!

I'd also like if we can have as a list in Extras menu, similar to how you have "in Trakt List"
(one cool idea is why not have all the groups displayed and an orange tick to show which you have as default)
so users can see alternative episode groups are available.
and a button in extras to.

one of the benefits of discovering this is that you might have specials in one group and not other like here

this is from Fenlight
![Screenshot (513)](https://github.com/user-attachments/assets/cd54d44f-8d14-4394-84da-117546accf6e)
 
and TMDB is using this
![Screenshot (514)](https://github.com/user-attachments/assets/b279dbd1-8a4a-4f78-acfc-ae20e5d9615f)

and I just realised TMDB already has this feature (I didn't check properly before)
![Screenshot (515)](https://github.com/user-attachments/assets/a9604cdf-b6d4-4691-b53a-c1a8ce610e12)
![Screenshot (516)](https://github.com/user-attachments/assets/0e49aeb1-88bf-473c-8cde-f2e2d224e732)

So now we pretty much have all anime come up properly, and Fenlight will be more awesome!
Hope you'll consider this, would be great and it's not just for anime like we've seen, some shows have different groups
thanks for your consideration

EDIT: I actually like TMDBhelpers method to, have the episode group inside the seasons

**Comment by hcgiub001** [comment]
Hey Peter
I did some more testing, seems as though the episode groups still have same naming for the season and episode for each listing

take "I woke up a vampire" for example. TMDB has it as 1 season 16 episode.
![Screenshot 2024-09-03 094534](https://github.com/user-attachments/assets/f17bc85f-920e-484e-b17c-562dc04ff65d)
when I change to the other group (Netflix Season (digital) )
it has 2 season, 8 episodes each
![Screenshot 2024-09-03 094721](https://github.com/user-attachments/assets/6599b93c-0070-489d-9e32-b1b2a64dd3b4)

but if you click on second season
![Screenshot 2024-09-03 094819](https://github.com/user-attachments/assets/3a0e4341-edad-43e8-aea0-97ee6660604c)

each episode at the far right has the same naming as first group (none netflix)
so S02E01 is 1x9, not 2x1 which is what we want

I checked with TMDBhelper and it's true
![Screenshot (519)](https://github.com/user-attachments/assets/b1f79ae8-e5f0-4e19-bfb2-476175aa6077)

TMDBhelper is just an informational addon so he has less use of passing the right information to the scrapers. But Fenlight does. I wonder if you can use the episode groups to know it's second season and find out what position the episode is in the season and pass that to the scraper?

kind of excited about this feature if you add it. plus would actually be useful beyond what TMDBhelper is doing.

even if you can't get the correct episode labelling, we can still make use of the information. Example

in "I woke up a vampire" you can go to season 2 and the first episode would be 2x1 etc
hope your well
thanks

**Comment by Tikipeter** [comment]
I'll look into it and see if they're consistently wrong and what can be done about it.

**Comment by hcgiub001** [comment]
Thankyou that's awesome! hope you can work your magic

**Comment by hcgiub001** [comment]
it's not that they are wrong. what I believe is happening is that TMDB have a naming structure. so each episode is linked to a metadata.
these "episode groups" I believe are created by the community.
all they do is create a new structure using TMDB's metadata. 
so it makes sense that TMDB points the episode back to their own structure.
but for our purposes we want to see the new episode naming within seasons. not the old one. This isn't something TMDBhelper dev thought of or cared about since it's just showing information and not using it to find episode links. we will. thanks

**Comment by Tikipeter** [comment]
Each episode has a unique identifier which it carries through all the various data of these different "groups". That's what's needed to write something that can make this possibly usable.

I'll delve deeper when I've got more time.

**Comment by hcgiub001** [comment]
I'll wait when you are ready, if that takes months then so be it.

**Comment by Tikipeter** [comment]
There's some good and bad news on the episode groups front.

The bad news is I really can't implement episode groups in a similar way to TMDbHelper. It's just not really feasible for an addon like Fen Light. There are too many moving pieces with relation to metadata.

The good news is I can utilize the episode group information to change which season/episode values are used when scraping. Which is really where their value lies for Fen Light.

For the next update, there will be a new option within Playback Options: "Scrape with Custom Episodes Group Value". If you choose this, you will be presented with the various episode groups that are available for that tvshow. If you select one of them, then Fen Light will attempt to match the current episode with the info, and return new season/episode values if possible.

For example, Bleach S01E121 within Fen Light will return no results, but if we use "Scrape with Custom Episodes Group Value", choose "TVDb Order" from the available groups, then the scrapers will instead search using S06E12 and we do then get results.

I will only include this implementation in the next update. If you'd like to test it a bit and let me know if it has value, then I'll implement a way to assign a certain group to a certain tvshow permanently.

Process:
Compare a normal scrape with using the episode group info would be good info.

Step 1 - Choose this:
![screenshot00006](https://github.com/user-attachments/assets/40df1af2-f925-4cda-ac86-778dcc95f5a9)

Step 2 - Choose this:
![screenshot00001](https://github.com/user-attachments/assets/7f369903-0781-4141-bff6-96faf8aa5d2f)

Step 3 - Select 1 of these to use for the scrape values:
![screenshot00002](https://github.com/user-attachments/assets/4fa895aa-5792-4e66-9627-e02a4b2f9fc3)

Just a general check to see if it works well enough to implement a more automatic approach to using the group info. The info for the groups are `GROUP NAME | GROUP TYPE | NUMBER OF EPISODES IN THE GROUP`.

There are 7 different types of groups you can get from TMDb, if someone has added them:
![Capture](https://github.com/user-attachments/assets/753b1151-32ce-4937-ae03-0c2c99ac97f0)

**Comment by Tikipeter** [comment]
Moved conversation to here.

**Comment by hcgiub001** [comment]
https://www.reddit.com/r/Addons4Kodi/comments/1d4qli1/need_help_with_how_fen_is_organizing_seasons/
Fen Lite default: 1 season 47 episodes
25 episodes onwards brings no result

Using episodes group season 51 episodes, works every time
![image](https://github.com/user-attachments/assets/b6966a83-f3d3-416a-ac57-0dbaaffe0845)
would be better if we could see how many seasons and episodes like tmdb shows
![image](https://github.com/user-attachments/assets/d1014818-b1f9-49b0-b095-b1711195f605)
better gauge which episode groups to pick

One Piece
https://www.reddit.com/r/Addons4Kodi/comments/1ela4bt/does_fenfen_lite_or_umbrella_provides_latest/

working perfectly for the latest season, the season before it It's not perfect

https://www.reddit.com/r/Addons4Kodi/comments/1f0dt9m/tmdb_season_metadata_option_for_shows_in_fen/

This show again works perfectly

I'm trying to find other animes that are like this, can't really find others.
It still would be great if we can pick the default episode grouping. One piece is the number 1 anime ever, so plenty would love to get sources for that.

Also I've seen you add a search for anime, great and thanks. One thing I realised is if I search for an anime, it doesn't store the search term in that section but in the tv-shows search section.

**Comment by hcgiub001** [comment]
yes I do, it almost makes the  "scrape with custom values" redundant for some use cases, like the ones those members asked for.

One thing I've heard but can't really find is we can actually create our own episode groups (or at least alter the other ones). So if an episode isn't perfect we can alter and see the changes in addons, this gives people motivations to help make tmdb better.

Great news about Comet scraper! hopefully you smart Dev's get it working, can't wait to see it! Again thanks for everything you do!

**Comment by hcgiub001** [comment]
I found two more series that wasn't working but did after picking the right episode groups
https://www.themoviedb.org/tv/65942-re
https://www.themoviedb.org/tv/62715-dragon-ball-chou/episode_group/607d7b58e6d3cc00775bab01

I've noticed if it's only one big season and nothing else and it doesn't load for later episodes, theirs a good chance it's not the right episode group.

With this , you're anime section in Fen Lite, more scrapers (Comet), I really think Fen Lite could work for most anime :D

and sometimes netflix does this where it picks it's own episode group that messes things up so this feature would be handy I think. thanks

**Comment by hcgiub001** [comment]
new series of Bleach anime is out and Fen Lite isn't able to find the new episode but TVDB order for episodes group , I get like 6 results to pick from.

This really does prove that we can access pretty much 99% of anime with this episode grouping and it seems the content is out their from Real debrid.

**Comment by hcgiub001** [comment]
https://www.themoviedb.org/tv/154526-mf/season/1
https://www.themoviedb.org/tv/131041
https://www.themoviedb.org/tv/210879?language=en-GB

3 more that using Episode Groups we can get links. I honestly think in terms of getting links General Addons can compete with dedicated addons with episode groups. The anime scene is getting more standardised like movie scene. any anime coming out nowadays is more of a matter that no one uploaded than the naming structure.

I'm really excited what this means for anime community having it integrated with Fen Light. 

Peter if I use these filter in TMDB airing today TV shows
![image](https://github.com/user-attachments/assets/c4b2d251-21d6-4f36-8c9f-fe051d314b0f)

I'm able to get anime that airs today. You could even choose other days by adjusting date
Haru addon has this
![image](https://github.com/user-attachments/assets/19ee75fb-dddd-42b3-b7c1-d3b97778ee32)

do you have api to do that for each day of this week? Monday to Sunday?
would be pretty amazing and the anime section really come in it's own. If not it's okay. the episode groups is so fun to play with and getting links for like 99% of anime now! :)

**Comment by hcgiub001** [comment]
I've also noticed this for a while but not mentioned it. the network list in extras menu for anime is not showing. in english movies and shows you have
![image](https://github.com/user-attachments/assets/5f1abae4-6917-4bea-9979-9abacc0d7ddc)
which shows other shows in the network, if that can be looked at some time after the massive list of request you had , I would be ever grateful. I think I won't request anything now. I don't have anything else to think about, Fen Light is complete thanks to your great work!

**Comment by Tikipeter** [comment]
With some time on my hands, I'll be tackling this next.

Sticking just to the episode groups portion of this issue, there is now a menu in CONTEXT MENU->OPTIONS: "Assign an Episode Group to [name_of_currently highlighted_show]". There you can set an episode group for that TV Show. This part has been done.

Then, of course, when it's time to scrape for sources, the relevant season/episode info will come from the assigned episode group. I'll sort this part out tomorrow.

After that, the plan is to at least have an option you can enable where if no sources are found then you can scrape again and the first episode group in the group results will be automatically used. This would need to be timed before, in the middle of, or after the "Retry With All Scrapers" and "Ignore All Filters" no result type settings.

I'll then need to write up a "Tips for Use" entry explaining what this is about.

**Comment by hcgiub001** [comment]
Great work, never thought about re-scrape with default group. looking forward to testing this out.

**Comment by hcgiub001** [comment]
On this episode
![image](https://github.com/user-attachments/assets/49bcc212-86d3-47ef-91f0-e10a13dac32a)

it shows as season 2 episode 29 (2x29)
I had to change Episode groups to "TVDB Order (Absolute)"

now when I search for this episode I get
![image](https://github.com/user-attachments/assets/908d80f6-2c89-4cd8-be1f-6ef9e8f4947b)

I can see it's actually searching for season 17 episode 29 (17x29)

I find it a disconnect to search for something we don't know if it's correct. For example imagine you are up to date with a show and watch the latest episode,  you've already assigned a different Episode Group, how can you be sure it's the correct file? we don't know what it's actually meant to search and hence the file names are useless. 

would it be possible to change the naming part so in my example it will change from

2x29 | THE DARK ARM
Bleach 

to

17x29 | THE DARK ARM
Bleach

so all the episodes will have to be changed.

if not the information for what we are searching for could be written at the top of Extra menus, that could also work though I still don't like the redundant default numbers 

Thanks and glad your back in Reddit

**Comment by hcgiub001** [comment]
Here is a good example

![image](https://github.com/user-attachments/assets/9e4b81c6-f20f-4e65-baa1-fcdbfc5a03a0)

so by default it's second season episode 3
when I do a normal search I get this
![image](https://github.com/user-attachments/assets/d3a9c022-f64a-4be0-9b9a-7cac6906686d)
when I change episode groups to "Absolute order"
I get this
![image](https://github.com/user-attachments/assets/64830848-5b7b-4b1f-8a29-6e4a64a6d52c)

so it's season 1 episode 29 
the cached result just says 29, I wasn't sure if that is the file I want but looking back at previous season it finishes at 26 so it makes sense the absolute order will give me file 29

I was wondering if we change the episode groups and you are willing and able to change the numbering of the episodes (from 2x3 to 1x29) that we should have a way to show the numbers have changed, having a colour will interfere with skins
how about 
{1x29}
using curly brackets to denote it's a different set? lol
well that's if it's possible to change without spending to long on it

**Comment by hcgiub001** [comment]
this show
![image](https://github.com/user-attachments/assets/1cc992fc-20bf-45e9-87c0-fe771afdb03d)
has seasons 
but checking on torrent site nyaa.si
![image](https://github.com/user-attachments/assets/74bbdaa5-7ca5-4efe-8808-aa8817b91a5e)
it seems it's all one season.

so I wanted to see how easy it is to create my own Episode group to have in one season, and I did

![image](https://github.com/user-attachments/assets/45b0e24c-144c-459f-8f40-d32ec947d27d)

It took about 2 minutes to create, I cleared cache but it still isn't showing. I assume it's because it will take time.

Kind of crazy that Fen Light now has the building blocks to correct TMDB and we can have access to shows we didn't before because of mismatch.

this encourages people to help TMDB and make it better. 
I've always thought we need a wiki for Fen Light, to have guides on every little part, 95% of people just need Fen Light as their addon and having a wiki will solve most problems

**Comment by hcgiub001** [comment]
I created my own custom group, takes a little while to see it

![image](https://github.com/user-attachments/assets/3c014c0f-acdf-4a3f-b655-edc56a4cc50a)

This opens up possibilities to tailor metadata on a specific torrent if you want.
and the whole Kodi network benefits

I've even thought of the possibility of having a specific code name for an Episode Group the kodi community fixes , then for Fen Light to have some function that it will alert you when it scrapes that it's seen an episode group with that code name and if the user wants to see what episode group it is. if they are happy with it then Fen Light will assign that group.
atm it's fine but the possibilities are great!
once the word spreads of this feature and how we can create custom metadata/episode groups then it'll catch on.
Another possibilities is having some file that we add the fixes, users can report, some moderator checks if the proposed Episode group for a specific show seems to work better. if it does it gets added to the file.

People can chose to add the file in Fen Light setup.
This feature is turning out to be a good tool.
great work!

**Comment by hcgiub001** [comment]
Thank you so much, it's working amazingly , you always seem to add features I was thinking of like how you have a prompt to ask if a user wants to explore other episode groups! awesome update. I'll be testing this. :)

**Comment by hcgiub001** [comment]
I've tested with one piece, i woke up a vampire, mf ghost, jujusu kaisen starting life in another world
all worked perfectly.

I tested with auto. worked great! I need more testing on this option, it could pick a group group so I wouldn't try this
(thoughts on auto: It would be better if we get the information which episode group is being tested, 
having picked another episode group and found a result, I would have expected it to have that group be assigned, but even if it wasn't I don't know what group actually was picked, if someone is picked auto , they know what the feature is and if we got the wrong group then they would stop and reassign, if they are happy then they will leave it, which makes more sense to have it automatic on that group)

I tested blue lock on manual, worked good.

I tested against this site
https://www.miruro.tv/

so I had a reference to go by
so far it's working better than I thought, I didn't think you'll be able to test other groups so fast (automatically)

I also tried testing what subtitles is it picking, the new episode group labelling or the old. It seemed it was the old, have you thought about this? how this affects what Fen Light passes to a4ksubtitles?

It's 1:17am so I will test more tommorow. just so amazed that you got it to do it automatically! 
thanks
